contracts/contracts/strategies/Generalized4626Strategy.sol

```diff
@@ -0,0 +1,141 @@
+// SPDX-License-Identifier: agpl-3.0
+pragma solidity ^0.8.0;
+
+/**
+ * @title OETH Generalized 4626 Strategy
+ * @notice Investment strategy for vaults supporting ERC4626 
+ * @author Origin Protocol Inc
+ */
+import { IERC4626 } from "../../lib/openzeppelin/interfaces/IERC4626.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import { IERC20, InitializableAbstractStrategy } from "../utils/InitializableAbstractStrategy.sol";
+
+contract Generalized4626Strategy is InitializableAbstractStrategy {
+    using SafeERC20 for IERC20;
+
+    IERC20 shareToken;
+    IERC20 assetToken;
+
+    /**
+     * @dev Deposit assets by converting them to shares
+     * @param _asset Address of asset to deposit
+     * @param _amount Amount of asset to deposit
+     */
+    function deposit(address _asset, uint256 _amount)
+        external
+        override
+        onlyVault
+        nonReentrant
+    {
+        _deposit(_asset, _amount);
+    }
+
+    /**
+     * @dev Deposit assets by converting them to shares
+     * @param _asset Address of asset to deposit
+     * @param _amount Amount of asset to deposit
+     */
+    function _deposit(address _asset, uint256 _amount) internal {
+        require(_amount > 0, "Must deposit something");
+        require(_asset == address(assetToken), "Unexpected asset address");
+
+        IERC4626(platformAddress).deposit(_amount, address(this));
+        emit Deposit(_asset, address(shareToken), _amount);
+    }
+
+    /**
+     * @dev Deposit the entire balance of assetToken to gain shareToken
+     */
+    function depositAll() external override onlyVault nonReentrant {
+        uint256 balance = assetToken.balanceOf(address(this));
+        if (balance > 0) {
+            _deposit(address(assetToken), balance);
+        }
+    }
+
+    /**
+     * @dev Withdraw asset by burning shares
+     * @param _recipient Address to receive withdrawn asset
+     * @param _asset Address of asset to withdraw
+     * @param _amount Amount of asset to withdraw
+     */
+    function withdraw(
+        address _recipient,
+        address _asset,
+        uint256 _amount
+    ) external override onlyVault nonReentrant {
+        require(_amount > 0, "Must withdraw something");
+        require(_recipient != address(0), "Must specify recipient");
+        require(_asset == address(assetToken), "Unexpected asset address");
+
+        IERC4626(platformAddress).withdraw(_amount, _recipient, address(this));
```

Comment:
Does this allow you to always withdraw the amount listed by checkBalance?

Reply:
will add a test for this to confirm

Reply:
Yes it allows the withdrawal of the whole balance. Test here: 
https://github.com/OriginProtocol/origin-dollar/pull/1271/commits/6c490c66d709636ecf0ba1970fe580b03ef70797